core/src/main/scala/zio/config/ReadError.scala

```diff
@@ -1,28 +1,56 @@
 package zio.config
 
-sealed trait ReadError[A] extends Exception { self =>
-  override def toString: String = self match {
-    case ReadError.MissingValue(path, message)          => s"MissingValue(${path}, ${message})"
-    case ReadError.FormatError(path, message)           => s"FormatError(${path}, ${message})"
-    case ReadError.ConversionError(path, message)       => s"ConversionError(${path},${message}"
-    case ReadError.OrErrors(list)                       => s"OrErrors(${list.map(_.toString)})"
-    case ReadError.AndErrors(list)                      => s"AndErrors(${list.map(_.toString)}"
-    case ReadError.ForceSeverity(error, treatAsMissing) => s"ForceSeverity(${error.toString},${treatAsMissing})"
-  }
+import zio.config.AnnotatedRead.Annotation
+import zio.config.ReadError.Step
+
+import scala.util.control.NoStackTrace
+
+sealed trait ReadError[A] extends Exception with NoStackTrace { self =>
+  def annotations: Set[Annotation]
+
+  final def fold[B](alternative: B)(f: PartialFunction[ReadError[A], B])(g: (B, B) => B, zero: B): B = {
+    def go(list: List[ReadError[A]]): B =
+      list.foldLeft(zero)((a, b) => g(b.fold(alternative)(f)(g, zero), a))
 
-  def size: Int =
     self match {
-      case ReadError.MissingValue(_, _)      => 1
-      case ReadError.FormatError(_, _)       => 1
-      case ReadError.ConversionError(_, _)   => 1
-      case ReadError.OrErrors(list)          => list.map(_.size).sum
-      case ReadError.AndErrors(list)         => list.map(_.size).sum
-      case ReadError.ForceSeverity(error, _) => error.size
+      case e @ ReadError.MissingValue(_, _, _)    => f.applyOrElse(e, (_: ReadError[A]) => alternative)
+      case e @ ReadError.FormatError(_, _, _, _)  => f.applyOrElse(e, (_: ReadError[A]) => alternative)
+      case e @ ReadError.ConversionError(_, _, _) => f.applyOrElse(e, (_: ReadError[A]) => alternative)
+      case e @ ReadError.Irrecoverable(list, _)   => f.applyOrElse(e, (_: ReadError[A]) => go(list))
+      case e @ ReadError.OrErrors(list, _)        => f.applyOrElse(e, (_: ReadError[A]) => go(list))
+      case e @ ReadError.ZipErrors(list, _)       => f.applyOrElse(e, (_: ReadError[A]) => go(list))
+      case e @ ReadError.ListErrors(list, _)      => f.applyOrElse(e, (_: ReadError[A]) => go(list))
+      case e @ ReadError.MapErrors(list, _)       => f.applyOrElse(e, (_: ReadError[A]) => go(list))
```

Comment:
We need to apply the function at each node, before it falls back to alternative or applying fold to each value in the list of errors.

With `alternative` not being present, `case e @ ReadError.Irrecoverable(list, _)  => false` will never pass

Reply:
Ok. In this case, I might pull it out of `ReadError`, since it is not likely to be general-purpose.